### PR TITLE
Feat/ll 72

### PR DESCRIPTION
**Pull request by oleksandrtarapata**
Completed tasks:

[LL-72](https://www.notion.so/limitlesslabs/7e5052e4a3a6414da3d2b6ddf8679519?v=587ac07e53b9425085bbe9eeb7832606&p=4072ccb9b7794a828d44a7fe1827813e&pm=s)

[LL-76](https://www.notion.so/limitlesslabs/7e5052e4a3a6414da3d2b6ddf8679519?v=587ac07e53b9425085bbe9eeb7832606&p=a4f0255949884f289a34d741c691ea83&pm=s)

### CODE REVIEW

File: src/app/wallet/components/WithdrawModal.tsx [hunk 1]
```diff
@@ -30,13 +43,47 @@ export const WithdrawModal = ({ onClose, isOpen, ...props }: Omit<IModal, 'child
 
   const disclosure = useDisclosure()
 
+  const tokenName = useMemo(() => {
+    return (
+      collateralTokensArray.find(
+        (collateralToken) => collateralToken.address[defaultChain.id] === selectedToken
+      )?.symbol || '$HIGHER'
```
Comment (valentynkvasov) [inline src/app/wallet/components/WithdrawModal.tsx:50]: Could you give some context why only || $HIGHER dedicated?

**Comment by valentynkvasov** [comment]
Super huge PR, @oleksandrtarapata you have done great job!!!

File: src/services/BalanceService.tsx [hunk 2]
```diff
@@ -80,29 +89,153 @@ export const BalanceServiceProvider = ({ children }: PropsWithChildren) => {
...
+          name: onChain.name,
+          decimals: onChain.decimals,
+          value: newBalanceBIOnChain,
+          formatted: formatUnits(newBalanceBIOnChain, onChain.decimals),
+          image: onChain.imageURI,
+          contractAddress: onChain.address[defaultChain.id],
+          price: marketTokensPrices ? marketTokensPrices['onchain'].usd : 0,
+        },
+      ]
+
```
Comment (maxschnaider) [inline src/services/BalanceService.tsx:228]: collateralTokens.forEach

File: src/app/wallet/components/DepositInfo.tsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+          <b>Buy ETH</b> on Coinbase, Binance or any other exchange / fiat onramp that supports
+          withdrawals on Base.
+        </Text>
+      </HStack>
+      <Spacer h={'8px'} />
+
+      <HStack w={'full'} spacing={4}>
+        <Avatar name='2' size={'sm'} bg={'grey.100'} color={'font'} fontWeight={'bold'} />
+        <Text>
+          <b>Send it</b> to the address below & ensure to select Base as the network.
```
Comment (valentynkvasov) [inline src/app/wallet/components/DepositInfo.tsx:18]: let's add <b> Base </b>

File: src/app/wallet/components/DepositInfo.tsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { Avatar, HStack, Spacer, Text } from '@chakra-ui/react'
+
+export default function DepositInfo() {
+  return (
+    <>
+      <HStack w={'full'} spacing={4}>
+        <Avatar name='1' size={'sm'} bg={'grey.100'} color={'font'} fontWeight={'bold'} />
+        <Text>
+          <b>Buy ETH</b> on Coinbase, Binance or any other exchange / fiat onramp that supports
+          withdrawals on Base.
```
Comment (valentynkvasov) [inline src/app/wallet/components/DepositInfo.tsx:10]: <b> Base </b>

File: src/types/types.ts [hunk 5]
```diff
@@ -47,4 +48,22 @@ export type GetBalanceResult = {
...
+export enum MarketTokensIds {
+  DEGEN = 'degen-base',
+  ETH = 'ethereum',
+  HIGHER = 'higher',
+  MFER = 'mfercoin',
+  ONCHAIN = 'onchain',
+  REGEN = 'regen',
+}
+
+export type GetCoingeckoPricesResponse = Record<MarketTokensIds, CoingeckoPriceEntity>
```
Comment (maxschnaider) [inline src/types/types.ts:69]: I'm a fan of DDD, imo src/types isn't the best place for this data

File: src/services/MarketsService.ts
Comment (maxschnaider) [inline src/services/MarketsService.ts:1]: good shot, but irrelevant to LL-72

File: src/providers/PriceOracle.tsx [hunk 6]
```diff
@@ -37,22 +44,68 @@ const PriceOracleContext = createContext<PriceOracleContextType | undefined>(und
...
+    console.log(data)
+    return data['degen-base'].usd as number
+  }
+
   const { data: ethPrice } = useQuery({
     queryKey: ['ethPrice'],
     queryFn: fetchEthPrice,
     staleTime: 1000 * 60 * 5, // Data life span 60 seconds
-    refetchInterval: 1000 * 60 * 5, // Refetch data every 60 seconds
+    // refetchInterval: 1000 * 60 * 5, // Refetch data every 60 seconds
```
Comment (valentynkvasov) [inline src/providers/PriceOracle.tsx:80]: should we remove it or uncomment?
It helped to optimise the price fetch before

File: src/providers/PriceOracle.tsx [hunk 7]
```diff
@@ -37,22 +44,68 @@ const PriceOracleContext = createContext<PriceOracleContextType | undefined>(und
...
+    return data
+  }
+
+  // const fetchCoinsList = async () => {
+  //   const { data } = await axios.get('https://api.coingecko.com/api/v3/coins/list')
+  //   console.info(data.filter((item: any) => item.name.toLowerCase().includes('degen')))
+  //   return data
+  // }
+
+  const fetchDegenPrice = async () => {
```
Comment (valentynkvasov) [inline src/providers/PriceOracle.tsx:68]: Why do we dedicate Degen & WEHT but others fetch by batch requesst?

File: src/services/MarketsService.ts
Comment (oleksandrtarapata) [reply src/services/MarketsService.ts:1]: I made this query in order to get active markets for user he/she bets and display it in a table.

File: src/providers/PriceOracle.tsx (hunk 6 above)
Comment (oleksandrtarapata) [reply src/providers/PriceOracle.tsx:80]: I assume it was disabled for some purposes, will return it back

File: src/providers/PriceOracle.tsx (hunk 7 above)
Comment (oleksandrtarapata) [reply src/providers/PriceOracle.tsx:68]: Just a non-deleted func created previously. Will remove

File: src/services/BalanceService.tsx [hunk 8]
```diff
@@ -80,29 +89,153 @@ export const BalanceServiceProvider = ({ children }: PropsWithChildren) => {
...
+      if (newBalanceBIWeth < parseEther('0.000001')) {
+        newBalanceBIWeth = 0n
       }
 
-      const balanceResult: GetBalanceResult = {
-        symbol: collateralToken.symbol,
-        decimals: collateralToken.decimals,
-        value: newBalanceBI,
-        formatted: formatUnits(newBalanceBI, collateralToken.decimals),
+      let newBalanceBIDegen = (await degenContract.read.balanceOf([smartWalletAddress])) as bigint
```
Comment (valentynkvasov) [inline src/services/BalanceService.tsx:128]: We have a lot of atomic reads, is there API to make bulk read for several smartcontract addresses?
I'm sure we might have more tokens, 10+,20+ we need to think how to optimise the reads

**Review by valentynkvasov — APPROVED** [review]

File: src/services/BalanceService.tsx (hunk 8 above)
Comment (oleksandrtarapata) [reply src/services/BalanceService.tsx]: need to think about proper approach

**Pull request merged**